### PR TITLE
Accesspoint feature enhancement with factory reset

### PR DESCRIPTION
**Pull request by md-rafeek**
## Description
[80638] - Require module support to perform Factory Reset AP
[75560] - Provision A to Site not present doesn't return any error
[80674] - Unable to configure channel number with Radio Role Assignment set to "Auto" via Ansible.
[83940] - Updating Power Level via Ansible is unsuccessful

## Type of Change
- [x] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Documentation update

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [x] Tasks are idempotent (can be run multiple times without changing state)
- [ ] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [x] Playbooks are modular and reusable
- [x] Handlers are used for actions that need to run on change

## Documentation
- [x] All options and parameters are documented clearly.
- [x] Examples are provided and tested.
- [x] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/accesspoint_workflow_manager.py [hunk 1]
```diff
@@ -543,6 +543,33 @@
...
             required: False
+      factory_reset_aps:
+        description: |
+          Factory reset one or more access points (APs) identified by their MAC addresses, hostnames, or management IP addresses.
+          At least one of the following parameters is required:
+          - mac_addresses
+          - hostnames
+          - management_ip_addresses
+        type: dict
+        required: False
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:554]: false

File: plugins/modules/accesspoint_workflow_manager.py [hunk 2]
```diff
@@ -543,6 +543,33 @@
...
+        suboptions:
+          mac_addresses:
+            description: A list of MAC addresses used to identify the access points for factory reset.
+            type: list
+            elements: str
+            required: False
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:560]: false

File: plugins/modules/accesspoint_workflow_manager.py [hunk 3]
```diff
@@ -543,6 +543,33 @@
...
+          hostnames:
+            description: |
+                A list of hostnames used to identify the access points for factory reset.
+            type: list
+            elements: str
+            required: False
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:566]: false

File: plugins/modules/accesspoint_workflow_manager.py [hunk 4]
```diff
@@ -543,6 +543,33 @@
...
+          management_ip_addresses:
+            description: |
+                A list of management IP addresses used to identify the access points for factory reset.
+            type: list
+            elements: str
+            required: False
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:572]: false

File: plugins/modules/accesspoint_workflow_manager.py [hunk 5]
```diff
@@ -1624,26 +1671,27 @@ def validate_ap_config_parameters(self, ap_config):
...
+                ap_identity_type = list(reboot_reset_aps.keys())[0]  # Retrieve the type (mac, hostname, or IP)
+                if ap_identity_type and ap_identity_type in self.keymap and len(reboot_reset_aps.get(ap_identity_type)) > 0:
+                    for ap_identifier in reboot_reset_aps[ap_identity_type]:
+                        if ap_identity_type == "mac_addresses":
+                            mac_regex = re.compile(r'^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$')
+                            if not mac_regex.match(ap_identifier):
+                                errormsg.append("mac_addresses: Invalid MAC Address '{0}' in playbook.".format(
+                                    str(ap_identifier)))
+                                self.log("Invalid MAC address: {0}".format(ap_identifier), "ERROR")
+                        elif ap_identity_type == "management_ip_addresses" and (not self.is_valid_ipv4(ap_identifier) and
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:1686]: Can we move the invalid IP check inside as like in mac_addresses for better readability?

```
                elif ap_identity_type == "management_ip_addresses":
                    if not self.is_valid_ipv4(ap_identifier) and not self.is_valid_ipv6(ap_identifier):
                        errormsg.append("management_ip_addresses: Invalid Management IP Address '{0}' in playbook."
                                        .format(ap_identifier))
                        self.log("Invalid Management IP address for {0}: {1}".format(reboot_reset, ap_identifier), "ERROR")
                
```

File: plugins/modules/accesspoint_workflow_manager.py [hunk 6]
```diff
@@ -1624,26 +1671,27 @@ def validate_ap_config_parameters(self, ap_config):
...
+            if reboot_reset_aps is not None:
+                ap_identity_type = list(reboot_reset_aps.keys())[0]  # Retrieve the type (mac, hostname, or IP)
+                if ap_identity_type and ap_identity_type in self.keymap and len(reboot_reset_aps.get(ap_identity_type)) > 0:
+                    for ap_identifier in reboot_reset_aps[ap_identity_type]:
+                        if ap_identity_type == "mac_addresses":
+                            mac_regex = re.compile(r'^([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})$')
+                            if not mac_regex.match(ap_identifier):
+                                errormsg.append("mac_addresses: Invalid MAC Address '{0}' in playbook.".format(
+                                    str(ap_identifier)))
+                                self.log("Invalid MAC address: {0}".format(ap_identifier), "ERROR")
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:1685]: self.log("Invalid MAC address for {0}: {1}".format(reboot_reset, ap_identifier), "ERROR")

File: plugins/modules/accesspoint_workflow_manager.py [hunk 7]
```diff
@@ -1624,26 +1671,27 @@ def validate_ap_config_parameters(self, ap_config):
...
+                        elif ap_identity_type == "management_ip_addresses" and (not self.is_valid_ipv4(ap_identifier) and
+                                                                                not self.is_valid_ipv6(ap_identifier)):
+                            errormsg.append("management_ip_addresses: Invalid Management IP Address '{0}' in playbook."
+                                            .format(str(ap_identifier)))
+                            self.log("Invalid Management IP address: {0}".format(ap_identifier), "ERROR")
+                        elif ap_identity_type == "hostnames":
+                            param_spec = dict(type="str", length_max=32)
+                            validate_str(ap_identifier, param_spec, "hostnames", errormsg)
+                            self.log("Hostname validation for '{0}' completed.".format(ap_identifier), "DEBUG")
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:1694]: self.log("Hostname validation for '{0}' in {1} completed.".format(ap_identifier, reboot_reset), "DEBUG")

File: plugins/modules/accesspoint_workflow_manager.py [hunk 8]
```diff
@@ -3119,6 +3173,113 @@ def access_point_reboot_status(self, task_id):
...
+        try:
+            self.log('Resetting the list of APs: {0}'.format(self.pprint(ap_list)), "INFO")
+            response = self.dnac._exec(
+                family="wireless",
+                function='factory_reset_access_points',
+                op_modifies=True,
+                params={"apMacAddresses": ap_list, "keepStaticIPConfig": True},
+            )
+            self.log("Response from factory_reset_access_points: {0}".format(str(response.get("response"))), "INFO")
+
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:3196]: Can we handle the error message and return earlier.
```
    if not response or not isinstance(response, dict):
        self.msg = "Failed to receive a valid response from 'factory_reset_access_points' API."
        self.log(self.msg, "ERROR"
        self.status = "failed"
        self.module.fail_json(msg=self.msg)

    task_id = response.get("response", {}).get("taskId")
    resync_retry_count = int(self.payload.get("dnac_api_task_timeout"))
```

File: plugins/modules/accesspoint_workflow_manager.py [hunk 9]
```diff
@@ -3119,6 +3173,113 @@ def access_point_reboot_status(self, task_id):
...
+            if response and isinstance(response, dict):
+                task_id = response.get("response", {}).get("taskId")
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:3198]: Check the task_id and if not return..
```
    if not task_id:
        self.msg = "Failed to retrieve task id from 'factory_reset_access_points' API response."
        self.log(self.msg, "ERROR"
        self.status = "failed"
        self.module.fail_json(msg=self.msg)
```

File: plugins/modules/accesspoint_workflow_manager.py [hunk 10]
```diff
@@ -3119,6 +3173,113 @@ def access_point_reboot_status(self, task_id):
...
+        try:
+            response = self.dnac._exec(
+                family="wireless",
+                function='get_access_points_factory_reset_status',
+                params=task_id,
+            )
+            self.log("Response from ap reset status: {0}".format(self.pprint(response)), "INFO")
+            if response and isinstance(response[0], dict):
+                return response.get("response", {})[0]
+            error_msg = "Invalid response format or missing data in AP reset status."
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:3273]: add a new line before..

File: plugins/modules/accesspoint_workflow_manager.py [hunk 11]
```diff
@@ -3193,6 +3355,27 @@ def main():
     else:
         ccc_network.log("No reboot list found in validated configuration.", "WARNING")
 
+    # Below set of line related to the Access Point factroy reset function.
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:3358]: Do we need to include all of these statements within the main function? We intended to maintain the module structure throughout our workflow modules. However, this AP workflow manager module appears to be distinct from the others.

File: plugins/modules/accesspoint_workflow_manager.py [hunk 12]
```diff
@@ -3193,6 +3355,27 @@ def main():
     else:
         ccc_network.log("No reboot list found in validated configuration.", "WARNING")
```
Comment (madhansansel) [inline plugins/modules/accesspoint_workflow_manager.py:3356]: Why do we need to print warning message to the user?

Comment (md-rafeek) [reply plugins/modules/accesspoint_workflow_manager.py:3356]: removed the else part

File: plugins/modules/accesspoint_workflow_manager.py (hunk 11 above)
Comment (md-rafeek) [reply plugins/modules/accesspoint_workflow_manager.py:3358]: created new function and moved.

**Review by md-rafeek — COMMENTED** [review]
addressed all comments

**Review by madhansansel — APPROVED** [review]

**Pull request merged**